torch/_inductor/runtime/triton_heuristics.py

```diff
@@ -481,12 +481,41 @@ def _precompile_config(self, cfg: Config, warm_cache_only: bool):
                 raise
             binary._init_handles()
 
+        """
+        https://github.com/pytorch/pytorch/issues/115344
+
+        self.fn.constexprs doesn't properly deal with None args, so when we filter out
+        an arg in UserDefinedTritonKernel.codegen, we need to filter it here as well.
+        We also don't want to modify self.fn.
+
+        We know that we removed something from the signature if:
+            1. It's in compile_meta["constants"]
+            2. It isn't a constant we already know about
+                Note: The value of interest has already been added to compile_meta['constants'],
+                    so we use self.fn.constexprs instead.
+            3. It isn't in the compile_meta signature
+        """
+        self.arg_names = self.fn.arg_names
+        removed_from_sig = set(compile_meta["constants"].keys())
+        known_constants = {
+            arg for i, arg in enumerate(self.arg_names) if i in self.fn.constexprs
+        }
+        removed_from_sig = removed_from_sig.difference(known_constants)
+        removed_from_sig = removed_from_sig.difference(
+            set(compile_meta["signature"].keys())
+        )
+
         call_args = [
             arg
-            for i, arg in enumerate(self.fn.arg_names)
-            if i not in self.fn.constexprs
+            for i, arg in enumerate(self.arg_names)
+            if i not in self.fn.constexprs and arg not in removed_from_sig
```

Comment:
One thing I don't quite understand is why we can't use`compile_meta["constants"].keys()` instead of `removed_from_sig` here and below? Constants get baked in during `triton.compile`, so should neither be in `def_args` nor `call_args`? And our `None` arg is in the `compile_meta["constants"].keys()` by now?

Reply:
So the logic here is this:

Elsewhere (in `UserDefinedTritonKernel.codegen(...)`) we filter out the None arg from `raw_args` which was added in `define_user_defined_triton_kernel(...)`.

So now we have to generate the proper bindings in Python so that our Triton kernel can be called---we have to filter out the parameter that we removed from the function signature (i.e., `call_args` and `def_args`). Specifically, we remove `None` args that aren't marked as `tl.constexpr`.

`compile_meta["constants"].keys()` does have the constants that were baked in earlier (including the None arg which was added in `define_user_defined_triton_kernel(...)`). However, this results in the following checks in `call_args` and `def_args` not being strict enough (and breaking other tests).

For example:

In the test `KernelTests.test_triton_kernel_equal_to_1_arg_dynamic_False`, we can see the following intermediate state:
```python
(Pdb) compile_meta["constants"].keys()
dict_keys(['BLOCK_SIZE', 'half_n_elements'])
(Pdb) known_constants # Note: this is just self.fn.constexprs mapped to the arg names
{'BLOCK_SIZE'}
(Pdb) compile_meta["signature"].keys()
dict_keys(['in_ptr0', 'in_ptr1', 'out_ptr', 'half_n_elements'])
(Pdb)
```

If we just use `compile_meta["constants"].keys()` as a stand-in for `removed_from_sig` in the following code:
```python
call_args = [
    arg
    for i, arg in enumerate(self.fn.arg_names)
    if i not in self.fn.constexprs and arg not in removed_from_sig
]
```

Then this would result in `half_n_elements` being removed (breaking the test). This specific failure happens because equal_to_1 args are treated as constants in `define_user_defined_triton_kernel`---but not included in `self.fn.constexprs`. These args are in the function signature and we don't want to filter these out from `call_args` or `def_args`.

`removed_from_sig = compile_meta["constants"].keys() - compile_meta["signature"].keys()` is also insufficient, breaking many tests. This is because `removed_from_sig` can end up containing `tl.constexpr` if not explicitly filtered out.

Example: `KernelTests.test_triton_kernel_with_imported_symbol_with_custom_name`:

These are the correct values for `call_args` and `def_args`---we see that `BLOCK_SIZE` (i.e., `tl.constexpr`) is included in def_args:
```python
(Pdb) call_args
['in_ptr', 'out_ptr', 'n_elements']
(Pdb) def_args
['in_ptr', 'out_ptr', 'n_elements', 'BLOCK_SIZE']
```
If we don't filter out the `tl.constexpr` values then this breaks the test.


Lastly, `removed_from_sig = compile_meta["constants"].keys() - tl.constexpr` is also insufficient, because tests such as `CustomOpTests.test_autotune_unbacked`---also because of equal_to_1 args (which are in the signature and constants---but we don't want to mark as having been removed).


So in conclusion, the logic we want is that we want to find the None args that we just removed---which we can find by:
1. Take `compile_meta["constants"].keys()`---which contains the value we want
2. Filtering out known `tl.constexpr` (since this doesn't match 1, due to equal_to_1 args)
3. Filtering out args present in the function signature

`removed_from_sig = compile_meta["constants"].keys() - tl.constexpr - compile_meta["signature"].keys()`

Reply:
So in short here what we want to do here is identify which None arguments we just removed from the list---and make sure we don't include them in `call_args` or `def_args`.

There are three sources of information at play:
1. `compile_meta["constants"].keys()`
2. `self.fn.constexprs`
3. `compile_meta["signature"].keys()`

Using (1) alone is insufficient because it isn't a strict enough measure of which argument we just removed---for example equal_to_1 args get included here, but not as a `tl.constexpr`, so this breaks several tests.

Using (1) - (3) is not sufficient because the set difference could include constexprs, which can be present in (1).

Using (1) - (2) breaks things because of equal_to_1 args, since we aren't filtering those out either (just the `None` args). 

The combination of (1) - (2) - (3) ensures that we only include the exact values we filtered out in `removed_from_sig`---so that the logic in `call_args` and `def_args` is sound (i.e., this set tracks the exact values we filtered out earlier).

Reply:
Thanks for the detailed explanation, this makes sense.